catalyst/contrib/callbacks/knn.py

```diff
@@ -0,0 +1,218 @@
+                                    validation should be calculated.
+            metric_fn (str): one of `accuracy`, `precision`, `recall`, 
+                            `f1-score`, default is `f1-score`
+            knn_metric (str): look sklearn.neighbors.NearestNeighbors parameter
+            n_neighbors (int): number of neighbors, default is 5
+        """
+
+        super().__init__(CallbackOrder.Metric)
+        self.prefix = prefix
+        self.features_key = input_key
+        self.targets_key = output_key
+
+        self.num_classes = num_classes
+        self.class_names = class_names
+
+        self.cv_loader_names = [] \
+            if cv_loader_names is None else cv_loader_names
+
+        if metric_fn == "accuracy":
+            self.metric_fn = accuracy_score
+        elif metric_fn == "recall":
+            self.metric_fn = recall_score
+        elif metric_fn == "precision":
+            self.metric_fn = precision_score
+        elif metric_fn == "f1-score":
+            self.metric_fn = f1_score
+        else:
+            raise ValueError(f"Metric function with value `{metric_fn}`"
+                             f" not implemented")
+
+        self.knn_metric = knn_metric
+        self.n_neighbors = n_neighbors
+
+        self.n_folds = 1
+
+        self.reset_cache()
+        self.reset_sets()
+
+    def on_batch_end(self, state: RunnerState):
+        features: torch.Tensor = \
+            state.output[self.features_key].cpu().detach().numpy()
+        targets: torch.Tensor = \
+            state.input[self.targets_key].cpu().detach().numpy()
+
+        self.features.extend(features)
+        self.targets.extend(targets)
+
+    def on_loader_end(self, state: RunnerState):
+
+        self.features = np.stack(self.features)
+        self.targets = np.stack(self.targets)
+
+        if len(np.unique(self.targets)) > self.num_classes:
+            raise Warning("Targets has more classes than num_classes")
+
+        s = {
+            "values": self.features,
+            "labels": self.targets,
+        }
+
+        self.sets[state.loader_name] = s
+
+        y_true, y_pred = self._knn(s)
+
+        loader_values = state.metrics.epoch_values[state.loader_name]
+        if self.num_classes == 2:
+            loader_values[self.prefix] = \
+                self.metric_fn(y_true, y_pred, average="binary")
+        else:
+            values = self.metric_fn(y_true, y_pred, average=None)
+
+            loader_values[f"{self.prefix}"] = np.mean(values)
+            for i, value in enumerate(values):
+                postfix = self.class_names[i] \
+                    if self.class_names is not None else str(i)
+                loader_values[f"{self.prefix}/{postfix}"] = value
+
+        self.reset_cache()
+
+    def on_epoch_end(self, state: RunnerState):
+
+        for s in self.cv_loader_names:
+            if s in self.sets:
+                y_true, y_pred = self._knn(self.sets["train"], self.sets[s])
```

Comment:
there are exists cases when we don't have "train" loader-name (it can be named like "train_1" for example), so this should be parametrized

Reply:
Added this functionality in `cv_loader_names` parameter.